src/showingpreviously/cinemas/merlin.py

```diff
@@ -0,0 +1,118 @@
+import re
+import time
+
+from bs4 import BeautifulSoup
+from datetime import datetime
+import showingpreviously.requests as requests
+from showingpreviously.model import ChainArchiver, CinemaArchiverException, Chain, Cinema, Screen, Film, Showing
+from showingpreviously.consts import UK_TIMEZONE
+
+
+CINEMAS_INDEX_URL = 'https://www.merlincinemas.co.uk/'
+FILM_URL = 'https://helston.merlincinemas.co.uk/ajax/film/{film_id}'
+SCREEN_NAME_PATTERN = re.compile(r'Showing In&nbsp;(?P<screen_name>.+?) -')
+
+CHAIN = Chain('Merlin Cinemas')
+
+
+def get_response(url: str) -> requests.Response:
+    r = requests.get(url)
+    if r.status_code != 200:
+        raise CinemaArchiverException(f'Got status code {r.status_code} when fetching URL {url}')
+    return r
+
+
+def get_cinemas_as_dict() -> dict[str, Cinema]:
+    r = get_response(CINEMAS_INDEX_URL)
+    soup = BeautifulSoup(r.text, features='html.parser')
+    cinema_list = soup.find('div', {'class': 'welcome_choice'})
+    cinemas = {}
+    for cinema_link in cinema_list.find_all('a', {'href': True}):
+        cinema_name = cinema_link.text
+        cinema_url = cinema_link['href']
+        if 'regal-theatre' in cinema_url:
+            continue  # skip the duplicate theatre
+        cinemas[cinema_url] = Cinema(cinema_name, UK_TIMEZONE)
+    return cinemas
+
+
+def get_films_info(film_ids: [str]) -> dict[str, Film]:
+    films = {}
+    for film_id in film_ids:
+        url = FILM_URL.format(film_id=film_id)
+        r = get_response(url)
+        soup = BeautifulSoup(r.text, features='html.parser')
+        film_title = soup.find('h1').text
+        try:
+            released_tag = soup.find('div', {'class': 'released'})
+            released_date = released_tag.find('span').text
+            released_year = released_date[-4:]
+        except AttributeError:
+            released_year = '0'
+        films[film_id] = Film(film_title, released_year)
+    return films
+
+
+def get_date_and_screen(booking_url: str, delay_time: int) -> (datetime, Screen):
+    try:
+        r = get_response(booking_url)
+    except requests.exceptions.ConnectionError:
+        # admit-one.eu has annoying rate limiting. try again, with an increased back-off
+        delay_time += 10
+        time.sleep(delay_time)
+        return get_date_and_screen(booking_url, delay_time)
+    soup = BeautifulSoup(r.text, features='html.parser')
+    date = None
```

Comment:
Is there a concern that this date could remain `None` when `get_date_and_screen()` returns?

Reply:
Not overly, because if that happens then it's not found a date for the showing, which is a bigger concern.